homeassistant/components/sensor/time_date.py

```diff
@@ -79,10 +87,21 @@ def icon(self):
         else:
             return 'mdi:clock'
 
-    @asyncio.coroutine
-    def async_update(self):
-        """Get the latest data and updates the states."""
-        time_date = dt_util.utcnow()
+    def get_next_interval(self, now=None):
+        """Compute next time an update should occur."""
+        if self.type == 'date':
+            interval = 86400
+        elif self.type == 'beat':
+            interval = 86.4
+        else:
+            interval = 60
+        if now is None:
+            now = dt_util.utcnow()
+        next_timestamp = int(dt_util.as_timestamp(now))
```

Comment:
This intent of this code is to have the sensor update approximately at the time that the minute/date actually changes.
I had two specific cases in mind that a straightforward timedelta fails to address.
1) Home Assistant restart: we have no way of knowing how many seconds into a minute a sensor will be instantiated. an extreme example is Home Assistant being restarted at 2 PM local time for a date sensor. waiting a full 86400 seconds to update state means the sensor will read the previous day's setting until 2 PM.
2) Time skew as Home Assistant runs. because it takes a finite amount of time for this code to run (and the system is cooperative multitasking), blindly asking for a 60-second sleep interval forever means this function will be called late. by re-computing how much time is left in the interval, we will stay on track.